Add WordPress.com toolbar navigation module

vindl opened the pull request:
_**Note: Do not merge before this is rebased and retargeted against `feature/settings-overhaul` branch**_

Ref: https://github.com/Automattic/jetpack/issues/6254
~Test with: D4478-code~ (deployed)

Improves the experience for Jetpack authors when they've transferred their WordPress site from WordPress.com to a self-hosted plan by providing a similar experience to the front-end navigation on WordPress.com.

#### Testing instructions:

1. Navigate to `General` tab of Jetpack Settings.
2. Enable WordPress.com Toolbar module.

![toolbar](https://cloud.githubusercontent.com/assets/1182160/23401481/275ef690-fda8-11e6-944a-41bc7a811326.png)

3. Verify that the core admin bar is replaced with the new WordPress.com Toolbar.
4. The menus should look and work the same as they do on WordPress.com.

![masterbar](https://cloud.githubusercontent.com/assets/1182160/22947175/4908d6de-f2fa-11e6-81d1-1739535a1196.png)



If you run into any issues while testing, please report them here for easier tracking: https://github.com/Automattic/jetpack/milestone/75

roundhill commented:
Tried this out, great start!  Noticed a few things:

- [ ] It'd be good to refresh the page after you activate the module so that you see the masterbar straight away. Same for when you deactivate it and go back to the old bar.
- [ ] When you hover over notifications it doesn't show the `pointer` cursor like the other items.

jeherve commented:
Worth noting that we're currently redesigning the Settings UI:
https://github.com/Automattic/jetpack/pull/6299
https://github.com/Automattic/jetpack/projects/3

You'll probably need to take this into account for the Jetpack Admin part. @eliorivero will be able to tell you a lot more.

eliorivero commented:
Hi @vindl I'm going to close this since PRs to add new settings or enhancement has to be in the new settings UI project and created against `feature/settings-overhaul`.
I'll comment more on your original issue.

vindl commented:
@jeherve @eliorivero Thanks for letting me know, I'll check it out and see how to adapt this for new settings UI.

vindl commented:
@eliorivero when we decide where to place this in the new UI, I can make the changes and rebase and retarget it against `feature/settings-overhaul` branch, so I think there is no need to close the PR. Would it be a problem if we kept it open until then, since it would make sharing changes with the rest of my team easier?

eliorivero commented:
Sure, that's ok 👍

beaulebens commented:
This is looking pretty good! I definitely think we need to come up with a better way of presenting it in the settings UI than "Masterbar".

Following are some notes from just testing this "functionally":

I definitely agree it'd be good to refresh the page after enabling/disabling, since otherwise nothing seems to happen, so as a user, I'd guess it's "broken".

When a site is in [development mode](https://jetpack.com/support/development-mode/), Notifications are not available, so they don't appear in the toolbar. Maybe we should remove that block entirely in that case, so that things slide over to the right properly? Similarly, if I disable Notifications in Jetpack, it doesn't entirely reclaim that far-right position, so there's an additional gap there.

![screen shot 2017-02-19 at 7 32 39 pm](https://cloud.githubusercontent.com/assets/108942/23109889/3163bc86-f6da-11e6-9ef8-33be7f106dc5.png)

I got a PHP notice:

`Notice: Undefined index: product_name_short in .../modules/masterbar/masterbar.php on line 480`

Had some missing icons in the My Sites menu (maybe tied specifically to the fact this site was in Dev Mode?):

![screen shot 2017-02-19 at 7 33 37 pm](https://cloud.githubusercontent.com/assets/108942/23109911/59375e0c-f6da-11e6-8175-389fe0ac5db4.png)

Also note that the URL shown in that menu is incorrectly encoded (install WP in a sub-directory, and you'll need to make sure you convert `::` back to `/` for display purposes).

No dynamic items in the Reader menu? I expected to see my Lists + Tags there, as I do on WP.com.

Under the "Me" menu, I noticed that your "Sign Out" action is just signing me out of the current site. If this is supposed to be creating a more WordPress.com-like experience, should that log me out of WP.com as well? TBH I'm not 100% sure what a normal user would really expect there (or if they'd fully grok all the different things going on).

I noticed 2 things with the Notifications menu: one is that the animation of the unread indicator appears to be triggering every time my window loses/gains focus (so make sure you have an unread, then switch tabs away and back again, and the indicator will animate). The other is that the icon being used is actually different to the one on WP.com:

![screen shot 2017-02-19 at 7 44 37 pm](https://cloud.githubusercontent.com/assets/108942/23110108/e33cdaa4-f6db-11e6-9dfe-7f979f0f1601.png)
![screen shot 2017-02-19 at 7 44 42 pm](https://cloud.githubusercontent.com/assets/108942/23110109/e33e0b90-f6db-11e6-9c1d-a7395db73531.png)

vindl commented:
Thanks for the feedback @beaulebens! I'll add these issues to our milestone (in AT repo) and look into resolving them as soon as possible.

vindl commented:
> Maybe we should remove that block entirely in that case, so that things slide over to the right properly?

Either that, or we could also make sure to activate Notifications module after Masterbar is activated - otherwise the admin bar will not match the one on WordPress.com.

> Had some missing icons in the My Sites menu (maybe tied specifically to the fact this site was in Dev Mode?)

Not sure what's causing this. My site is in Dev mode too, and the icons are displayed correctly.

> No dynamic items in the Reader menu?

Haven't carried them over yet, will make sure to add them in future updates.

> If this is supposed to be creating a more WordPress.com-like experience, should that log me out of WP.com as well?

I suppose we should log them out of WordPress.com as well.

beaulebens commented:
> Note sure what's causing this
Is your site publicly accessible? The one I saw that on was on localhost, so maybe it was somehow related to not being accessible from external sites?

> log them out of WordPress.com
You might be able to achieve this with something like the `redirect_to` parameter (I think that's what it's called. So generate a local log out link, and then set the redirect to point to a WP.com log out. So they'd log out locally, get redirected, which would log them out of WP.com, and then get dropped on the log out screen from WP.com (the one talking about our apps). I think that'd work (as long as a nonce is not required to log out of WP.com).

vindl commented:
>Is your site publicly accessible? The one I saw that on was on localhost, so maybe it was somehow related to not being accessible from external sites?

Yup, it is. That's probably what's causing it.

roundhill commented:
Good idea to just pull the CSS from wp.com. We should add a README and/or some code comments to warn devs that changes made to those will also impact Jetpack.

vindl commented:
> We should add a README and/or some code comments in the wp.com files to warn devs that changes made to those will also impact Jetpack.

Good point @roundhill, I've added the comments in the diff linked above.

@beaulebens I've created issues based on your testing feedback [in this milestone](https://github.com/Automattic/jetpack/milestone/75) for easier tracking - most of them should be resolved by now. 

> No dynamic items in the Reader menu? I expected to see my Lists + Tags there, as I do on WP.com.

They are not currently present in WP.com toolbar either (only in Calypso), so I've decided to skip them for now.

eliorivero commented:
We need to add a support article to Jetpack.com site since right now the support link in this card goes to the right place `https://jetpack.com/support/masterbar/` and there's currently no article there.

roundhill commented on modules/module-headings.php:
```diff
@@ -91,6 +91,11 @@ function jetpack_get_module_i18n( $key ) {
 
+			'masterbar' => array(
+				'name' => _x( 'WordPress.com Toolbar', 'Module Name', 'jetpack' ),
+				'description' => _x( 'An admin bar providing similar experience to the front-end navigation on WordPress.com.', 'Module Description', 'jetpack' ),
```
How about:

`Replaces the admin bar with a useful toolbar to quickly manage your site via WordPress.com.`

or maybe:

`Adds a useful toolbar to quickly manage your site via WordPress.com.`

cc @drw158 for any additional copy ideas.

vindl replied:
Replaced it with the first suggestion in 0660ea26a58439a22995179e04b38be434d821f6.

jeherve commented on modules/masterbar/masterbar.php:
```diff
@@ -0,0 +1,821 @@
+
+		if ( $this->is_rtl() ) {
+			// Extend core WP_Admin_Bar class in order to add rtl styles
+			add_filter( 'wp_admin_bar_class', function () {
```
Could you remove this anonymous function, as this will create Fatal errors in PHP 5.2?

Thanks!

vindl replied:
Thanks for catching this! Updated in 0660ea26a58439a22995179e04b38be434d821f6.

jeherve commented on modules/masterbar/masterbar.php:
```diff
@@ -0,0 +1,821 @@
+		$i   = hexdec( substr( md5( $file ), - 1 ) ) % 2;
+		$url = 'http://s' . $i . '.wp.com' . $file;
+
+		return set_url_scheme( $url );
```
Since all the files are served from WordPress.com, maybe we can just always serve them via HTTPS? What's the advantage of serving them via HTTP?

vindl replied:
No advantage, it's just something that I've overlooked. Updated to use HTTPS in 0660ea26a58439a22995179e04b38be434d821f6.

jeherve commented on modules/masterbar/masterbar.php:
```diff
@@ -0,0 +1,821 @@
+		$wp_admin_bar->add_menu( array(
+			'parent' => 'newdash',
+			'id'     => 'streams-header',
+			'title'  => __( 'Streams', 'jetpack' ),
```
This might benefit from a translator comment, because "Streams" doesn't mean much and is hard to translate without context.

vindl replied:
Provided additional context in 0660ea26a58439a22995179e04b38be434d821f6, not sure if it's clear enough.

jeherve commented on modules/masterbar/masterbar.php:
```diff
@@ -0,0 +1,821 @@
+		$wp_admin_bar->add_menu( array(
+			'parent' => $id,
+			'id'     => 'special-header',
+			'title'  => __( 'Special', 'jetpack' ),
```
This might also benefit from a translator comment.

jeherve commented on modules/masterbar/masterbar.php:
```diff
@@ -0,0 +1,821 @@
+		wp_enqueue_script( 'a8c_wpcom_masterbar_overrides', $this->wpcom_static_url( '/wp-content/mu-plugins/admin-bar/masterbar-overrides/masterbar.js' ) );
+	}
+
+	public function get_site_slug() {
```
We have a similar function in `class.jetpack.php`, named `build_raw_urls`:
https://github.com/Automattic/jetpack/blob/master/class.jetpack.php#L6197-L6208

We usually use it like `Jetpack::build_raw_urls( get_home_url() );`.

Maybe we could use this for the Masterbar as well instead of creating a new function?

vindl replied:
Great! I wasn't aware of it. Refactored code to use `Jetpack::build_raw_urls()` instead in 0660ea26a58439a22995179e04b38be434d821f6.

jeherve commented on modules/masterbar/masterbar.php:
```diff
@@ -0,0 +1,821 @@
+			}
+		}
+
+		if ( current_user_can( 'manage_options' ) ) {
```
Since the links direct to WordPress.com, maybe we should only show this to admins whose accounts are linked to WordPress.com for that site? If their account isn't linked, the links won't work for them.

vindl replied:
In case their account is not connected we will bail in constructor, and the masterbar won't be shown at all (https://github.com/Automattic/jetpack/pull/6413/files#diff-d73a71fc4f0617de9989f4321bc4f39cR20).

jeherve commented on modules/notes.php:
```diff
@@ -104,7 +104,16 @@ function action_init() {
+		 *
+		 * @param bool true Should notifications be displayed in RTL style. Defaults to false.
+		 */
+		$enqueue_rtl = apply_filters( 'enqueue_rtl_notification_styles', false );
```
Should we prefix this filter to avoid potential conflicts with other plugins?

vindl replied:
Good point, prefixed with `a8c_wpcom_masterbar_` in 0660ea26a58439a22995179e04b38be434d821f6.

jeherve commented on modules/notes.php:
```diff
@@ -104,7 +104,16 @@ function action_init() {
+		/**
+		 * Can be used to force Notifications to display in RTL style.
+		 *
+		 * @since 4.7.0
```
4.7 is already in Beta, so that's most likely going to be 4.8.0

vindl replied:
Updated in 0660ea26a58439a22995179e04b38be434d821f6.

jeherve commented on modules/masterbar.php:
```diff
@@ -0,0 +1,16 @@
+
+include dirname( __FILE__ ) . '/masterbar/masterbar.php';
+
+new A8C_WPCOM_Masterbar;
```
Since we're loading quite a few files from WordPress.com, maybe we should prefetch the domains we use for those resources, like we do for other modules? Here is an example with the Comments module:
https://github.com/Automattic/jetpack/blob/master/modules/comments.php#L33-L42

vindl replied:
Thanks for suggesting this! Added domain prefetching in 0660ea26a58439a22995179e04b38be434d821f6.

jeherve commented on modules/masterbar.php:
```diff
@@ -0,0 +1,16 @@
+ * Module Description: An admin bar providing similar experience to the front-end navigation on WordPress.com.
+ * Sort Order: 38
+ * Recommendation Order: 16
+ * First Introduced: 4.7
```
This is most likely going to be 4.8 since 4.7 is already in Beta.

vindl replied:
Updated to `4.8` in 0660ea26a58439a22995179e04b38be434d821f6.

roundhill commented on modules/module-info.php:
```diff
@@ -335,6 +335,24 @@ function jetpack_custom_css_more_info() { ?>
+	esc_html_e(
+		"Improves the experience for Jetpack authors when they've transferred their WordPress " .
+		"site from WordPress.com to a self-hosted plan by providing a similar experience to the " .
+		"front-end navigation on WordPress.com."
```
Here's an idea for an improvement to this description:

`Quickly access your Stats, Notifications, Posts and more on WordPress.com. The Toolbar is displayed for any user on the site that is connected to WordPress.com.`

vindl replied:
Updated description as suggested in 0660ea26a58439a22995179e04b38be434d821f6.

eliorivero commented on modules/notes.php:
```diff
@@ -104,7 +104,16 @@ function action_init() {
-		if ( !is_rtl() ) {
+		/**
+		 * Can be used to force Notifications to display in RTL style.
+		 *
```
This needs
```
* @module notes
```
to be added.

vindl replied:
:+1: Added in 27a75daf1bf60dec6e43f0de4ab643a83c5dd451.

dereksmart commented on modules/masterbar/masterbar.php:
```diff
@@ -0,0 +1,823 @@
+				// On the frontend the `url` query arg will load that page in the Customizer and also return to it after closing
+				// non-home URLs won't work unless we undo domain mapping since the Customizer preview is unmapped to always have HTTPS
+				global $current_blog;
+				$current_page = 'https://' . $current_blog->domain . $_SERVER['REQUEST_URI'];
```
When viewing a single post/page, looks like $current_blog is not set as you expected.  I am getting 
`Notice: Trying to get property of non-object in /home/dereksma/public_html/wp-content/plugins/jetpack/modules/masterbar/masterbar.php on line 666` on single post views on the front end.

vindl replied:
Nice catch @dereksmart! I've resolved this in a1b9ab96713cdd7d215d8be34ed3a4aba27460a8.

dereksmart commented:
Other than the thing I pointed out above, it tests really well.  

Tested: 
- logged in/out of wordpress.com (no regressions from current behavior, i.e. notifications)
- Multiple user levels 
- Linked & unlinked users

ryelle commented:
I noticed some accessibility issues, so I created a separate PR for fixing those: #6598

rickybanister commented:
@MichaelArestad let's chat today about the settings for this—the settings in this PR match the old style, for it to be merged we'll need to update that and find the right place for it (we we briefly discussed yesterday).

jeherve commented:
I still have some testing to do, but I had a few remarks after playing with the Masterbar for a bit:

- When I first activated the module, the page refreshed and the masterbar appeared, but clicking "My Sites" directed me to WordPress.com instead of triggering the dropdown.
- Clicking "Switch Sites" leads me to WordPress.com Sites instead of allowing me to switch sites from my site.
- The "Domains" and the WP Admin menu items probably shouldn't be displayed in the left menu.
- If the Stats module is disabled on my site, I would expect the Stats menu item to disappear from the Masterbar menu.
- Same thing with the Sharing menu if I disable Jetpack's Publicize and Jetpack's Sharing.
- The Masterbar doesn't seem to pick up my Gravatar; it shows mysteryman instead.
- Should we edit the "Help" link and direct it to http://jetpack.com/support/ instead?

eliorivero commented:
- [ ] When I'm linked as an admin, I can see the Gravatar, but I can't when I'm linked as a non admin, an Editor in this case. In the split image, admin is on the left, editor on the right.

    <img width="342" alt="admin on the right, non admin on the left" src="https://cloud.githubusercontent.com/assets/1041600/23865575/e2441dea-07f4-11e7-8c0e-8d99fe038ad9.png">

    <img width="263" alt="captura de pantalla 2017-03-13 a las 14 05 09" src="https://cloud.githubusercontent.com/assets/1041600/23865959/1c2a9948-07f6-11e7-8ae0-ea0299d417e0.png">

- [ ] When I'm linked as a Subscriber, these links should be hidden since it looks like I can post in the site

    <img width="297" alt="captura de pantalla 2017-03-13 a las 14 01 26" src="https://cloud.githubusercontent.com/assets/1041600/23865828/aa449e28-07f5-11e7-8f8a-c9860a71751c.png">

    and when I click to add a post, I'm taken to

    <img width="1079" alt="captura de pantalla 2017-03-13 a las 14 02 32" src="https://cloud.githubusercontent.com/assets/1041600/23865855/c7c6cbec-07f5-11e7-95f9-c217c93fdfcc.png">

vindl commented:
@jeherve 

> When I first activated the module, the page refreshed and the masterbar appeared, but clicking "My Sites" directed me to WordPress.com instead of triggering the dropdown.

It looks like this was occurring in a short interval before all the required masterbar scripts were loaded. I've removed the URLs from top level menu items in 445241f1ea8cc5bfb47b3b4daee382064925d6bc in order to prevent this.

> Clicking "Switch Sites" leads me to WordPress.com Sites instead of allowing me to switch sites from my site.

The same is true for current version of WordPress.com masterbar. In this iteration our main goal is just to replicate the current functionality as closely as possible.

> The "Domains" and the WP Admin menu items probably shouldn't be displayed in the left menu.
> ...
> Should we edit the "Help" link and direct it to http://jetpack.com/support/ instead?

I think that for AT sites we should show these three menu items and preserve the original WordPress.com links. On the other hand, for non-AT Jetpack sites it would make sense to hide them, or direct them to corresponding Jetpack URLs. Do you think that adding such AT specific check to Jetpack would be appropriate?

> If the Stats module is disabled on my site, I would expect the Stats menu item to disappear from the Masterbar menu.

Fixed in b2c32e86ccd688a5e110ae237a271d589c52554d. 

> Same thing with the Sharing menu if I disable Jetpack's Publicize and Jetpack's Sharing.

Fixed in b2c32e86ccd688a5e110ae237a271d589c52554d.

> The Masterbar doesn't seem to pick up my Gravatar; it shows mysteryman instead.

Resolved in 51ddbd53c9d9180ed8f12ce8c87f50c1a416958b.

@eliorivero 

>When I'm linked as an admin, I can see the Gravatar, but I can't when I'm linked as a non admin, an Editor in this case. In the split image, admin is on the left, editor on the right.

Resolved in 51ddbd53c9d9180ed8f12ce8c87f50c1a416958b.

> When I'm linked as a Subscriber, these links should be hidden since it looks like I can post in the site.

Fixed in 01ef071248fe737c01f9f75e5d3302d49e596ed4.


Thanks a lot for testing and feedback @jeherve and @eliorivero!

beaulebens commented:
Note that the settings overhaul is being merged today, so this is going to need an update to match/fit into that style of settings now.

rickybanister commented:
cc @ashleighaxios in case we can provide any design help on updating the Masterbar feature's settings card now that the new settings has merged (with Michael afk early next week).

dereksmart commented:
This is looking good enough to merge.  The smaller things will be addressed in following PR's, as will the merge to the new settings branch.
